src/modules/avpops/avpops.c

```diff
@@ -1049,9 +1051,36 @@ static int w_op_avps(struct sip_msg* msg, char* param, char *op)
 								(struct fis_param*)op);
 }
 
-static int w_subst(struct sip_msg* msg, char* src, char *subst)
+static int w_subst(struct sip_msg* msg, char* src, char *subst_param)
 {
-	return ops_subst(msg, (struct fis_param**)src, (struct subst_expr*)subst);
+	str subst;
+	struct subst_expr* se;
+	fparam_t *fp;
+
+	fp = (fparam_t*)subst_param;
+
+	if (fp->type == FPARAM_SUBST) // already precompiled subst
+	{
+		se = fp->v.subst;
+		return ops_subst(msg, (struct fis_param**)src, se);
+	}
+
+	if (get_str_fparam(&subst, msg, fp) != 0)
+	{
+		LM_ERR("error fetching subst re");
+		return -1;
+	}
+
+	LM_DBG("%s preparing %s\n", exports.name, subst.s);
+
+	se=subst_parser(&subst);
+	if (se==0)
+	{
+		LM_ERR("%s: bad subst re %s\n", exports.name, subst.s);
+		return E_BAD_RE;
+	}
+
+	return ops_subst(msg, (struct fis_param**)src, se);
```

Comment:
I think that se needs to be freed, because subst_parser() allocates the result.

Reply:
Yes, indeed. Thanks!